include/envoy/matcher/matcher.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/config/common/matcher/v3/matcher.pb.h"
+#include "envoy/config/typed_config.h"
+
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+#include "external/envoy_api/envoy/config/core/v3/extension.pb.h"
+
+namespace Envoy {
+namespace Matcher {
+
+template <class DataType> class MatchTree;
+
+template <class DataType> using MatchTreeSharedPtr = std::shared_ptr<MatchTree<DataType>>;
+
+using TypedExtensionConfigOpt = absl::optional<envoy::config::core::v3::TypedExtensionConfig>;
+
+// On match, we either return the action to perform or another match tree to match against.
+template <class DataType> struct OnMatch {
+  const TypedExtensionConfigOpt action_;
+  const MatchTreeSharedPtr<DataType> matcher_;
+};
+
+/**
+ * MatchTree provides the interface for performing matches against the data provided by DataType.
+ */
+template <class DataType> class MatchTree {
+public:
+  virtual ~MatchTree() = default;
+
+  // This encodes a three states:
+  // Not enough data to complete the match: {false, {}}
+  // Completed the match, no match: {true, {}}
+  // Completed the match, match: {true, on_match}
+  struct MatchResult {
+    const bool match_completed_;
+    const absl::optional<OnMatch<DataType>> on_match_;
+  };
+
+  // Attempts to match against the matching data (which should contain all the data requested via
+  // matching requirements). If the match couldn't be completed, {false, {}} will be returned.
+  // If a match result was determined, {true, action} will be returned. If a match result was
+  // determined to be no match, {true, {}} will be returned.
+  virtual MatchResult match(const DataType& matching_data) PURE;
+};
+
+// InputMatcher provides the interface for determining whether an input value matches.
+class InputMatcher {
+public:
+  virtual ~InputMatcher() = default;
+
+  /**
+   * whether the provided input is a match.
+   * @param absl::optional<absl::string_view> the value to match on. Will be absl::nullopt if the
+   * lookup failed.
+   */
+  virtual bool match(absl::optional<absl::string_view> input) PURE;
+};
+
+using InputMatcherPtr = std::unique_ptr<InputMatcher>;
+
+/**
+ * Factory for registering custom input matchers.
+ */
+class InputMatcherFactory : public Config::TypedFactory {
+public:
+  virtual InputMatcherPtr createInputMatcher(Protobuf::Message& config) PURE;
+
+  std::string category() const override { return "envoy.matching.matcher"; }
```

Comment:
```suggestion
  std::string category() const override { return "envoy.matching.input_matcher"; }
```

? We might have other extensions later.